runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -4039,11 +4107,11 @@ ssize_t wrap_fi_send(c_nodeid_t node,
     DBG_DO_PRINTF("tx AM send to %d: %s, ctx %p",
                   (int) node, am_reqStr(node, req, reqSize), ctx);
   }
+  tcip->numTxnsOut++;
```

Comment:
Can you say why you're moving these? Is it in some way required by the new logic in here that the counters be updated before we fire the transactions? I don't object to it on principle, especially if it's now needed for correctness, but when I originally coded all this I felt like we shouldn't count chickens before we'd hatched them, so to speak.

Reply:
If these variables aren't  incremented before initiating the operation (e.g. `fi_send`) then the routines that check for completion such as `checkTxCmplsCntr` and functions that invoke it such as `waitForTxnComplete` may think the operation has completed when it has not.  Most of the time the check is in a loop, e.g. this code in `waitForTxnComplete`:
```
    while (tcip->numTxnsOut > 0) {
      sched_yield();
      (*tcip->ensureProgressFn)(tcip);
    }
```
so I'm not sure it's an issue, but not all of them are in a loop. Furthermore, when I added debugging statements the completion messages appeared before the `libfabric` messages indicating the operation had been initiated, which is confusing at best. 

Another way to put it, is if they are incremented after `OFI_RIDE_OUT_EAGAIN` they are incorrect within `OFI_RIDE_OUT_EAGAIN` which checks for completion as a part of ensuring progress.

Reply:
Initially my response to the above was something like "Holy Toledo! `OFI_RIDE_OUT_EAGAIN()` is the source of a _lot_ of unnecessary overhead!".  🤯     On further reflection I don't think it's quite that bad, but still this seems like an area where I didn't think about all the ramifications as the comm layer evolved, and where we could do better. I'll follow up with an issue. In the meantime, I do now understand that this and the related changes elsewhere are right.

Reply:
I've gone and re-read the man page text about `FI_EAGAIN`, and now I'm back to thinking this may not be right. What `OFI_RIDE_OUT_EAGAIN()` does is to call the give function until it doesn't return `FI_EAGAIN`. According to the `fi_msg(3)` man page, which is where this is documented and where all the other applicable man pages point, functions return `FI_EAGAIN` when they can't get local resources to initiate operations. When this happens it's due to previously-initiated operations, and the caller needs to ensure progress for previous operations only (because they're the problem) in order to free up the associated resources so the current call can proceed. We don't need to progress the current call, and certainly don't need to wait for it to complete, because if we get `FI_EAGAIN` it was not initiated.

If you saw debug output appearing in the wrong order there must have been something else going on. With respect to

> ... when I added debugging statements the completion messages appeared before the libfabric messages indicating the operation had been initiated, which is confusing at best.

what "libfabric messages indicating the operation had been initiated" are you talking about? Output from within libfabric itself, as a result of running with the `FI_LOG_LEVEL` env var set?

Reply:
Ignore the AM handler for now as the same basic logic applies. `OFI_RIDE_OUT_EAGAIN` calls the given function, e.g. `fi_send`. Assume it returns `FI_EAGAIN`. `OFI_RIDE_OUT_EAGAIN` then calls `tcip->ensureProgressFn` which is set to `tcip->checkTxCmplsFn` which is set to `checkTxCmplsCntr`. `checkTxCmplsCntr` compares the current counter value with `numTxnsSent` and sets `numTxnsOut` to the difference between the two. I added a debug message that printed when `numTxnsOut` equaled zero because that indicated that all outstanding transmissions have completed. But with the original control flow `numTxnsOut` and `numTxnsSent` are not set until after `OFI_RIDE_OUT_EAGAIN`, so my debug message would print before the transmission completed. That's because `checkTxCmplsCntr` is using variables that have yet to be updated. When I turned on libfabric debugging by setting `FI_LOG_LEVEL=Debug` I would see my debug message even before the `libfabric` message indicating the transmission had been initiated. 

Looking at the code again I realize that if the transmission were to complete between the call to `fi_send` and reading the counter in `checkTxCmplsCntr`, the `if (count > tcip->numTxnsSent)` conditional would be true and it would abort. That never happened although if it had it would have been easier to figure out what was going on.

Reply:
The `FI_EAGAIN` return is a purely local thing, though. If a libfabric function returns that, it means that insufficient resources were available to initiate the communication that function would have done, and thus the comm did not happen. Any counter values when that happens have to do only with previously-initiated operations, not the one we're trying to fire. (In fact, depending on what resources were lacking, you might even expect that the function wouldn't stop returning `FI_EAGAIN` until after the counters "caught up".)

> When I turned on libfabric debugging by setting FI_LOG_LEVEL=Debug I would see my debug message even before the libfabric message indicating the transmission had been initiated.

Right, that matches: libfabric will only have initiated the transmission if the function _doesn't_ return `FI_EAGAIN`. If your debug output is done for counter catch-up after the function did return `FI_EAGAIN`, it's just because all previous comms are done, not this one.

Reply:
The function `checkTxCmplsCntr` reads the variable `numTxnsSent` and uses it to set `numTxnsOut`. This function is invoked by `OFI_RIDE_OUT_EAGAIN` if `FI_EAGAIN` is returned by `libfabric`.  If `write_fi_send` modifies `numTxnsSent` after calling `OFI_RIDE_OUT_EAGAIN` then `checkTxCmplsCntr` will at best read the unmodified value of the variable, and possibly incorrectly set `numTxnsOut` to zero if the counter hasn't incremented, or at worst call `INTERNAL_ERROR_V` if the counter has incremented because it is greater than `numTxnsSent`.

Reply:
But if a libfabric function returns `FI_EAGAIN` it didn't initiate an operation, and `fi_cntr_read()` will certainly never read a counter value indicating completion of an un-initiated operation. So I don't see how that counter could become greater than an un-incremented `numTxnsSent` in `checkTxCmplsCntr()`.

Reply:
Greg's analysis was correct and I fixed my code.